src/app/interceptors/custom-header.interceptor.ts

```diff
@@ -0,0 +1,40 @@
+/**
+ *    Copyright 2020 OICR
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+import { HttpEvent, HttpHandler, HttpInterceptor, HttpRequest } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { Observable } from 'rxjs';
+import { v4 as uuidv4 } from 'uuid';
+import { versions } from '../footer/versions';
+
+const uiVersion = versions.tag;
+const uuid = uuidv4().toString();
+
+/**
+ * An interceptor that ensures that every request has a custom X-Dockstore-UI header with the UI version as
+ * the value and a custom UUID header so we can track API calls coming from the Dockstore UI.
+ */
+@Injectable()
+export class CustomHeaderInterceptor implements HttpInterceptor {
+  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+    req = req.clone({
+      setHeaders: {
+        'X-Dockstore-UI': uiVersion,
+        'X-UUID': uuid
```

Comment:
1. I prefer something like X-Dockstore-UUID, or maybe [X-Request-Id](https://stackoverflow.com/questions/25433258/what-is-the-x-request-id-http-header). But don't feel too strongly about it.
2. Alternatively, you could generate a different uuid per request, but that might be slow -- this is definitely a major improvement as is.

Reply:
Ideally (in a world where we don't care about privacy), we would want one UUID per browser. So we know what a person is up to and what problems they encounter. 
In a more realistic world, one UUID per browser session would be enough to preserve privacy but let us know what the user path through the site looks like. 
And having said that, I guess you (@coverbeck ) were more thinking about tracing single REST requests as they made their way through the LB, nginx, and webservice weren't you ..

Reply:
> And having said that, I guess you (@coverbeck ) were more thinking about tracing single REST requests as they made their way through the LB, nginx, and webservice weren't you ..

Yes, I was thinking of that case.

Reply:
The way I implemented it, there would be one UUID per browser session. If the user goes to Dockstore in a new tab, or refreshed Dockstore on their current tab, a new UUID would be generated.

@coverbeck If you want, I could add a X-Request-ID header as well and generate a UUID for each request? In which case, @denis-yuen would you still want a X-Dockstore-UUID header that has a UUID for each browser session?

Reply:
Yes, I think the X-Request-ID header would be a good idea and can/should co-exist with the UUID header you have already added. Thanks!

Reply:
Despite my lobbying for it, I am slightly concerned there might be a performance hit generating a UUID on every request. Can you please at least informally measure how long it takes to generate one? Thanks!

Reply:
So I used performance.now() before and after the call that generates a UUID and calculated the difference to get an idea of how long it takes.

I logged the results in the console:
![Screen Shot 2020-04-02 at 12 41 26 PM](https://user-images.githubusercontent.com/25287123/78275417-95eba980-74df-11ea-82b7-e203825853bc.png)

Looks like it's roughly 0.03 - 0.05 ms to generate a UUID.

Reply:
Neat and TIL: `performance.now()` -- I'd always used `Date.now()`. :)

But performance.now() returns a value in milliseconds, so how are you calculating that it completes in less than a millisecond? Is is actually 3-5 milliseconds? Probably OK if so, but as a further data point, are you running on a powerful machine?

Reply:
I'm just doing ``endTime - startTime`` :) 

According to [this](https://developers.google.com/web/updates/2012/08/When-milliseconds-are-not-enough-performance-now), ``performance.now()`` is a double with microseconds in the fractional, so that's probably why the elapsed time is less than a millisecond. Just a note in case you use ``performance.now()`` in the future, [this](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now) says that the precision of ``performance.now()`` is also dependent on browser settings, so for example Firefox would round it to the closest millisecond by default (I used Chrome and it didn't round).

I don't think I'm running on a particularly powerful machine. Here are the specs if you want to judge for yourself:
![Screen Shot 2020-04-02 at 1 19 36 PM](https://user-images.githubusercontent.com/25287123/78279234-3b554c00-74e5-11ea-8680-0049f37065a1.png)

Reply:
Ah yes, I was trying it in FF and reading [this](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now).

Reply:
So a UUID for each request should be okay then?

Reply:
I'm OK with it.